src/offset/local/windows.rs

```diff
@@ -8,158 +8,222 @@
-///
-/// # Safety
-///
-/// The provided error ID must align with the provided Windows API, providing the wrong ID could lead to UB.
-macro_rules! windows_sys_call {
-    ($name:ident($($arg:expr),*), $error_id:expr) => {
-        if $name($($arg),*) == $error_id {
-            return Err(Error::last_os_error());
-        }
-    }
-}
-
-const HECTONANOSECS_IN_SEC: i64 = 10_000_000;
-const HECTONANOSEC_TO_UNIX_EPOCH: i64 = 11_644_473_600 * HECTONANOSECS_IN_SEC;
+use super::{FixedOffset, Local, TzInfo};
+use crate::{DateTime, Datelike, LocalResult, NaiveDate, NaiveDateTime, NaiveTime, Weekday};
 
+// We don't use `GetLocalTime` because its results can be ambiguous, while conversion from UTC
+// never fails.
 pub(super) fn now() -> DateTime<Local> {
-    LocalSysTime::local().datetime()
+    let mut now = MaybeUninit::<SYSTEMTIME>::uninit();
+    unsafe { GetSystemTime(now.as_mut_ptr()) }
+    // SAFETY: GetSystemTime cannot fail according to spec, so we can assume the value
+    // is initialized.
+    let now_naive = systemtime_to_naive_dt(unsafe { now.assume_init() }, 0).unwrap();
+
+    naive_to_local(&now_naive, false).single().expect("Current local time must exist")
 }
 
-/// Converts a local `NaiveDateTime` to the `time::Timespec`.
 pub(super) fn naive_to_local(d: &NaiveDateTime, local: bool) -> LocalResult<DateTime<Local>> {
-    let naive_sys_time = system_time_from_naive_date_time(d);
-
-    let local_sys_time = match local {
-        false => LocalSysTime::from_utc_time(naive_sys_time),
-        true => LocalSysTime::from_local_time(naive_sys_time),
+    let tz_info = match TzInfo::get_current_for_year(d.year()) {
+        Some(i) => i,
+        None => return LocalResult::None,
     };
-
-    if let Ok(local) = local_sys_time {
-        return LocalResult::Single(local.datetime());
+    match local {
+        false => tz_info.local_from_utc_time(d),
+        true => tz_info.local_from_local_time(d),
     }
-    LocalResult::None
 }
 
-struct LocalSysTime {
-    inner: SYSTEMTIME,
-    offset: i32,
-}
-
-impl LocalSysTime {
-    fn local() -> Self {
-        let mut now = MaybeUninit::<SYSTEMTIME>::uninit();
-        unsafe { GetLocalTime(now.as_mut_ptr()) }
-        // SAFETY: GetLocalTime cannot fail according to spec, so we can assume the value
-        // is initialized.
-        let st = unsafe { now.assume_init() };
-
-        Self::from_local_time(st).expect("Current local time must exist")
+// The basis for Windows timezone and DST support has been in place since Windows 2000. It does not
+// allow for complex rules like the IANA timezone database:
+// - A timezone has the same base offset the whole year.
+// - There are either zero or two DST transitions.
+// - As of Vista(?) only years from 2004 until a few years into the future are supported.
+// - All other years get the base settings, which seem to be that of the current year.
+impl TzInfo {
+    fn get_current_for_year(year: i32) -> Option<Self> {
```

Comment:
So this would still cause windows to return `LocalResult::None` in cases outside of the 1601-30827 range, right?

Is it possible to adjust for handling dates outside of that range so that the behavior is stable whether it is used on linux or windows.

Reply:
That is what the comment below it is about :smile:. The Windows API has those limitations, so we clamp the year for values this far into the past or future. Whatever timezone rules are extrapolated for these years (and Windows extrapolates it really naive...) is what can be extrapolated for years beyond. But the clamped `ref_year` is only used for getting the timezone info, not for calculating the local datetime.

Reply:
Ah, sorry. I really should've double checked `clamp` (100% my bad). 

In this case though, since we can be certain that `ref_year` should always be valid, would it be better to actually panic when `GetTimeZoneInformationForYear` returns 0 instead of returning `None` and hiding the error?

Reply:
I hope that in the future when chrono uses `Result`s instead of `Option`s for fallible operations this can return something like `Err(ChronoError::LocalTzLoadingError)`. Currently `None`, which gets converted to `LocalResult::None`, is the only way to return an error, similar to the Unix implementation.

Reply:
My thought was that you could panic instead of returning `None`, which might be the better option in this scenario, In theory, this should never fail from an incorrect parameter error, so the source of the error in that case may warrant panicking instead of swallowing the error with `None`.

Reply:
Totally agree with that. I was looking for other errors and happened upon someone mentioning that another syscall in `GetTimeZoneInformationForYear` can throw an oom error, but other than that I've found no reference to this call being fallible if the years are valid.

Also maybe this [doc](https://github.com/MicrosoftDocs/sdk-api/blob/docs/sdk-api-src/content/timezoneapi/nf-timezoneapi-gettimezoneinformationforyear.md) might be helpful